imitation_learning.md

```diff
@@ -0,0 +1,50 @@
+# Imitation Learning
+
+We will use imitation learning to train a policy that controls the
+boat. Imitation learning is the study of algorithms that improve
+performance in making decisions by observing demonstrations from a
+teacher[1].
+
+
+Our algorithm will be very similar to DAGGER [2]. The core idea of
+DAGGER is to train the first policy from the teacher's demonstration,
+and then let the policy explore the environment, the teacher only
+corrects the policy after the exploration in hindsight. Allowing the
+trained policy to explore the environment makes the training data
+distribution closer to the inference distribution, rather than the
+teacher demonstration's data distribution. Significantly improve the
+reliability of the trained policy.
+
+## DAGGER Pseudo Algorithm
+
+1. The teacher demonstrates the task, collect the states, and the teacher's actions.
+1. Initialize the dataset as the list of collected pairs of state and action.
+1. for i = 1 to N do # Train the policy on the dataset.
+   1. Use the trained policy, collect some states.
+   1. In hindsight, ask the teacher the action for the collected state.
+   1. Append the newly collected pairs of state and action to the dataset.
+   1. Train the policy on the dataset.
+
+## Our Pseudo Algorithm
+
+In the boat navigation case, the teacher is a person. The problem with
+applying DAGGER directly to driving a boat is that it's hard to give
+the sensor data to the teacher and ask him what action to take - the
+teacher has to be on the boat in the exact moment to know what action
+to take. So we modify the DAGGER algorithm to:
+
+1. The teacher demonstrates driving the boat, collect the states, and the teacher's actions.
+1. Initialize the dataset as the list of collected pairs of state and action.
+1. for i = 1 to N do # Train the policy on the dataset.
+   1. Let the boat self-drive with the trained policy.
+   1. The teacher can take over the boat's control at any time and drive with the teacher's policy, collect the pairs of state and action as D0.
+   1. Append D0 to the dataset.
+   1. When the teacher is driving, train the new policy with the new dataset in real-time.
+
+We think that training the new policy in real-time is essential as it
```

Comment:
Sure, but it isn't the most mandatory thing if it introduces technical challenges that make it counter productive

Reply:
good point, let's adjust this design doc once we have more information (e.g., know the technical challenges that make it counter productive).